fdbclient/DatabaseContext.h

```diff
@@ -329,6 +329,9 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
 	// map from tssid -> metrics for that tss pair
 	std::unordered_map<UID, Reference<TSSMetrics>> tssMetrics;
 
+	// map from ssid -> ss tag
+	std::unordered_map<UID, Tag> ssidTagMapping;
```

Comment:
I think the code would be simpler if you instead change the `locationCache` to include the tag. I understand this might not be trivial, but it would simplify a lot of the code in `NativeAPI`

Reply:
I am not sure if it is logical to include the tag info in locationCache. Locationcache is mapping key range to locationInfo so not sure if the corresponding ssid -> tag map should be a part of that.

Reply:
`LocationInfo` is just a list of `StorageServerInterface` instances. Each of them are associated with one tag.

Reply:
I just realized: one difficulty here will be to pass the correct version to `loadBalance` when you send the request. I think a simple `std::min` would do the trick, but my guess is that it could potentially lead to more `transaction_too_old` errors. So a better solution would be to plumb this somehow into `loadBalance` (which is not exactly a pretty piece of code).

Reply:
I currently took the approach of including the latest commit versions of all storage server replicas (that serve the key) in the client request - this avoids changing the load balancer code and the receiver will still find its latest commit version in the request but at the cost of including "replication_factor" number of latest commit versions in the request (and probably higher if a split/merge is in progress), but probably this overhead is not that bad.

Reply:
I think that works as well -- so in practice you will ship between 3 and 6 tags to the storage, that doesn't sound like a meaningful overhead.

I still would prefer to get rid of the above structure though. If you store the tag in the `StorageServerInterface` you can just fetch the tags from your cache and keeping it up to date will be free (for your implementation)

Reply:
Here is my attempt at populating store tag in StorageServerInterface class:

```
diff --git a/fdbclient/StorageServerInterface.h b/fdbclient/StorageServerInterface.h
index bddca6fcb..9c5fe5496 100644
--- a/fdbclient/StorageServerInterface.h
+++ b/fdbclient/StorageServerInterface.h
@@ -58,6 +58,7 @@ struct StorageServerInterface {
        LocalityData locality;
        UID uniqueID;
        Optional<UID> tssPairID;
+       Tag tag;
 
        RequestStream<struct GetValueRequest> getValue;
        RequestStream<struct GetKeyRequest> getKey;
diff --git a/fdbserver/storageserver.actor.cpp b/fdbserver/storageserver.actor.cpp
index fe964eb6c..26a9c638c 100644
--- a/fdbserver/storageserver.actor.cpp
+++ b/fdbserver/storageserver.actor.cpp
@@ -5148,7 +5148,7 @@ ACTOR Future<Void> memoryStoreRecover(IKeyValueStore* store, Reference<ClusterCo
 
 // for creating a new storage server
 ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
-                                 StorageServerInterface ssi,
+                                 StorageServerInterface& ssi,
                                  Tag seedTag,
                                  Version tssSeedVersion,
                                  ReplyPromise<InitializeStorageReply> recruitReply,
@@ -5181,6 +5181,8 @@ ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
                        self.tag = seedTag;
                }
 
+               ssi.tag = self.tag;
+
                self.storage.makeNewStorageServerDurable();
                wait(self.storage.commit());
 
@@ -5208,7 +5210,7 @@ ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
        }
 }
 
-ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface ssi) {
+ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface& ssi) {
        ASSERT(!ssi.isTss());
        state Transaction tr(self->cx);
 
@@ -5271,6 +5273,7 @@ ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface
                                                                self->tag = rep.tag;
                                                        }
                                                        self->allHistory = self->history;
+                                                       ssi.tag = self->tag;
 
                                                        TraceEvent("SSTag", self->thisServerID).detail("MyTag", self->tag.toString());
                                                        for (auto it : self->history) {
@@ -5299,7 +5302,7 @@ ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface
        return Void();
 }
 
-ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterface ssi) {
+ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterface& ssi) {
        // RYW for KeyBackedMap
        state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(self->cx);
        state KeyBackedMap<UID, UID> tssMapDB = KeyBackedMap<UID, UID>(tssMappingKeys.begin);
@@ -5333,6 +5336,7 @@ ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterfa
 
                        wait(tr->commit());
                        self->tag = myTag;
+                       ssi.tag = self->tag;
 
                        break;
                } catch (Error& e) {


```

Reply:
However, I am nervous about making this change (I am not even confident that this change is all that we need to do this). Maybe I will keep the code that I have for now and revisit this later?

Reply:
This looks correct to me. There should only be these two places where this is getting written. I would expect this approach to be less error prone than your current one.

Reply:
Thanks for going through the diff! I will apply this diff (and drop the ssid -> tag map) at some later point.